### ISSUE TITLE
Compile error during Linux build in Docker container.

### ISSUE THREAD

**Issue by NotChristianGarcia** [microdee/UE4-SpaceMouse#102]
Hello, I'm new to Unreal, trying to build this plugin.

- [I'm using the guide on building for linux here.](https://github.com/microdee/UE4-SpaceMouse/blob/872788c3a250ad1dcc9b40a2936d144e232ec44c/LINUX.md) 
- I'm building in the official docker container
  - `docker run --rm -ti -v "<UE4-SpaceMouseDir>:/UE4-SpaceMouse" ghcr.io/epicgames/unreal-engine:dev-5.3.2`
  - I tried 5.3.1 and 5.2 as well
- Once in container I'm installing libudev-dev + extras 
  - `sudo apt update; sudo apt install libudev-dev g++ gdb make ninja-build rsync zip gcc libhidapi-dev -y`
- Then I'm running the build script
  - `~/UnrealEngine/Engine/Build/BatchFiles/RunUAT.sh BuildPlugin -Plugin="/UE4-SpaceMouse/SpaceMouse.uplugin" -Package="/UE4-SpaceMouse/Out/Linux"`
- I'm on nixos, thus building in Docker. @eotsn if you have a way out of this grave of mine, that would be appreciated. 

I'm however running into the following identical error for `[2/9] Compile Module.HIDUE.cpp`, `[3/9] Compile Module.SpaceMouseRuntime.cpp`, `[4/9] Compile Module.SpaceMouseReader.cpp`, and `[7/9] Compile Module.SpaceMouseEditor.cpp`. I've only included the first for brevity. 

I'm out of my depth with DotNet and unreal building in general. I could imagine this being a dependency issue maybe? I saw that one user installed VS 2022 workloads, don't have that in linux though (please correct me if I'm wrong and I'll give it a shot). Asking in case anyone had wisdom on this. 

```
ue4@dockerew:~/UnrealEngine$ ~/UnrealEngine/Engine/Build/BatchFiles/RunUAT.sh BuildPlugin -Plugin="/UE4-SpaceMouse/SpaceMouse.uplugin" -Package="/UE4-SpaceMouse/Out/Linux" 
                                                                                                                                                                                
Running AutomationTool...                                                                                                                                                       
                                                                                                                                                                                
Fixing inconsistent case in filenames.                                                                                                                                          
Setting up Mono                                                                                                                                                                 
Setting up bundled DotNet SDK                                                                                                                                                   
Start UAT Interactively: dotnet AutomationTool.dll BuildPlugin -Plugin=/UE4-SpaceMouse/SpaceMouse.uplugin -Package=/UE4-SpaceMouse/Out/Linux                                    
/home/ue4/UnrealEngine/Engine/Build/BatchFiles/Linux/../../../Binaries/ThirdParty/DotNet/6.0.302/linux/dotnet                                                                   
Starting AutomationTool...                                                                                                                                                      
Parsing command line: BuildPlugin -Plugin=/UE4-SpaceMouse/SpaceMouse.uplugin -Package=/UE4-SpaceMouse/Out/Linux                                                                 
Initializing script modules...                                                                                                                                                  
Total script module initialization time: 0.15 s.                                                                                                                                
Executing commands...                                                                                                                                                           
Copying 161 file(s) using max 64 thread(s)                                                                                                                                      
Reading plugin from /UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/SpaceMouse.uplugin...                                                                              
Building plugin for host platforms: Linux                                                                                                                                       
Running: /home/ue4/UnrealEngine/Engine/Binaries/ThirdParty/DotNet/6.0.302/linux/dotnet "/home/ue4/UnrealEngine/Engine/Binaries/DotNET/UnrealBuildTool/UnrealBuildTool.dll" Unrea
lEditor Linux Development -Project="/UE4-SpaceMouse/Out/Linux/HostProject/HostProject.uproject" -plugin="/UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/SpaceMouse.upl
ugin" -noubtmakefiles -manifest="/UE4-SpaceMouse/Out/Linux/HostProject/Saved/Manifest-UnrealEditor-Linux-Development.xml" -nohotreload -log="/home/ue4/Library/Logs/Unreal Engin
e/LocalBuildLogs/UBT-UnrealEditor-Linux-Development.txt"                                                                                                                        
Log file: /home/ue4/Library/Logs/Unreal Engine/LocalBuildLogs/UBT-UnrealEditor-Linux-Development.txt                                                                            
/UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/Source/HIDUE/HIDUE.Build.cs: warning: Referenced directory '/usr/lib/aarch64-linux-gnu' does not exist.                
/UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/Source/HIDUE/HIDUE.Build.cs: warning: Referenced directory '/usr/local/lib/x86_64-linux-gnu' does not exist.           
/UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/Source/HIDUE/HIDUE.Build.cs: warning: Referenced directory '/usr/local/lib/aarch64-linux-gnu' does not exist.          
Parsing headers for UnrealEditor                                                                                                                                                
  Running Internal UnrealHeaderTool /UE4-SpaceMouse/Out/Linux/HostProject/HostProject.uproject /UE4-SpaceMouse/Out/Linux/HostProject/Intermediate/Build/Linux/UnrealEditor/Devel
opment/UnrealEditor.uhtmanifest -WarningsAsErrors -installed                                                                                                                    
Total of 16 written                                                                                                                                                     
Reflection code generated for UnrealEditor in 3.8202378 seconds                                                                                                         
------- Build details --------                                                                                                                                          
Using toolchain located at '/home/ue4/UnrealEngine/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v22_clang-16.0.6-centos7/x86_64-unknown-linux-gnu'.           
Using clang (/home/ue4/UnrealEngine/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v22_clang-16.0.6-centos7/x86_64-unknown-linux-gnu/bin/clang++) version 'clang version
 16.0.6 (github.com/llvm/llvm-project 7cbf1a2591520c2491aa35339f227775f4d3adf6)                                                                                         
Target: x86_64-unknown-linux-gnu                                                                                                                                        
Thread model: posix                                                                                                                                                     
InstalledDir: /home/ue4/UnrealEngine/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v22_clang-16.0.6-centos7/x86_64-unknown-linux-gnu/bin' (string), 16 (major), 0 (mino
r), 6 (patch)                                                                                                                                                           
Using bundled libc++ standard C++ library.                                                                                                                              
Using lld linker                                                                                                                                                        
Using llvm-ar (/home/ue4/UnrealEngine/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v22_clang-16.0.6-centos7/x86_64-unknown-linux-gnu/bin/llvm-ar) version 'LLVM (http:
//llvm.org/):                                                                                                                                                           
  LLVM version 16.0.6                                                                                                                                                   
  Optimized build. (string)'                                                                                                                                            
Using fast way to relink  circularly dependent libraries (no FixDeps).                                                                                                  
------------------------------                                                                                                                                          
Writing manifest to /UE4-SpaceMouse/Out/Linux/HostProject/Saved/Manifest-UnrealEditor-Linux-Development.xml                                                              
Building UnrealEditor...                                                                                                                                                
Determining max actions to execute in parallel (8 physical cores, 8 logical cores)                                                                                    
  Executing up to 16 processes, one per physical core                                                                                                                   
  Requested 1.5 GB memory per action, 5 GB available: limiting max parallel actions to 3                                                                             
------ Building 9 action(s) started ------                                                                                                                              
[1/9] Compile Module.HIDUE.cpp                                                                                                                                          
In file included from /UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/Intermediate/Build/Linux/x64/UnrealEditor/Development/HIDUE/Module.HIDUE.cpp:2:          
In file included from /UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/Source/HIDUE/Private/HIDUE.cpp:4:                                                         
In file included from /UE4-SpaceMouse/Out/Linux/HostProject/Plugins/SpaceMouse/Source/HIDUE/Public/HIDUE.h:6:                                                            
In file included from Runtime/Core/Public/Modules/ModuleManager.h:5:                                                                                                    
In file included from Runtime/Core/Public/Containers/Array.h:7:                                                                                                         
In file included from Runtime/Core/Public/Misc/ReverseIterate.h:6:                                                                                                      
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/iterator:684:                                                                                               
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/__iterator/common_iterator.h:22:                                                                             
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/variant:218:                                                                                                
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/tuple:227:                                                                                                  
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/utility:251:                                                                                                
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/compare:145:                                                                                                
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/__compare/compare_partial_order_fallback.h:13:                                                               
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/__compare/partial_order.h:14:                                                                               
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/__compare/weak_order.h:14:                                                                                  
In file included from ThirdParty/Unix/LibCxx/include/c++/v1/__compare/strong_order.h:18:                                                                                
/home/ue4/UnrealEngine/Engine/Source/ThirdParty/Unix/LibCxx/include/c++/v1/cmath:544:5: error: use of undeclared identifier '__promote'                                  
    __promote<_A1, _A2, _A3>                                                                                                                                            
    ^                                                                                                                                                                   
/home/ue4/UnrealEngine/Engine/Source/ThirdParty/Unix/LibCxx/include/c++/v1/cmath:548:22: error: no template named '__promote'                                            
    typedef typename __promote<_A1, _A2, _A3>::type __result_type;                                                                                                      
                     ^                                                                                                                                                  
/home/ue4/UnrealEngine/Engine/Source/ThirdParty/Unix/LibCxx/include/c++/v1/cmath:552:12: error: call to 'hypot' is ambiguous                                             
    return hypot((__result_type)__lcpp_x, (__result_type)__lcpp_y, (__result_type)__lcpp_z);                                                                             
           ^~~~~                                                                                                                                                        
/home/ue4/UnrealEngine/Engine/Source/ThirdParty/Unix/LibCxx/include/c++/v1/cmath:533:46: note: candidate function                                                        
inline _LIBCPP_INLINE_VISIBILITY float       hypot(       float __x,       float __y,       float __z ) { return sqrt(__x*__x + __y*__y + __z*__z); }                   
                                             ^                                                                                                                          
/home/ue4/UnrealEngine/Engine/Source/ThirdParty/Unix/LibCxx/include/c++/v1/cmath:534:46: note: candidate function                                                        
inline _LIBCPP_INLINE_VISIBILITY double      hypot(      double __x,      double __y,      double __z ) { return sqrt(__x*__x + __y*__y + __z*__z); }                   
                                             ^                                                                                                                          
/home/ue4/UnrealEngine/Engine/Source/ThirdParty/Unix/LibCxx/include/c++/v1/cmath:535:46: note: candidate function                                                        
inline _LIBCPP_INLINE_VISIBILITY long double hypot( long double __x, long double __y, long double __z ) { return sqrt(__x*__x + __y*__y + __z*__z); }                   

...
shortened
...

3 errors generated.
[8/9] Link (lld) libUnrealEditor-SpaceMouseEditor.so cancelled
[9/9] WriteMetadata UnrealEditor.target cancelled
Took 21.967016599999997s to run dotnet, ExitCode=6
UnrealBuildTool failed. See log for more details. (/home/ue4/Library/Logs/Unreal Engine/LocalBuildLogs/UBT-UnrealEditor-Linux-Development.txt)
AutomationTool executed for 0h 0m 22s
AutomationTool exiting with ExitCode=6 (6)
RunUAT ERROR: AutomationTool was unable to run successfully. Exited with code: 6

```

**Comment by eotsn** [comment]
I vaguely remember having similar issues when compiling with the `build-essential` meta-package installed on my system. My solution then was to uninstall the package, but it seems like it's part of the image so that's not an option to test out :thinking: 

I never tried to build with Docker so I don't really have any advice to give... Since you're on Nix an option could be to put together an environment using `nix-shell` and a `shell.nix` file and use the binary release of Unreal Engine. Then you'd have full control of the build environment.

**Comment by microdee** [comment]
I'd also try to compile and use the plugin in the project you're working in rather than using UAT to package a plugin. Plugin packaging is one of those things vague and undocumented in Unreal (other than it exists). If you still want to package this plugin on Linux, then try it with another plugin to see if it's problem with UE4-Spacemouse itself or engine/your environment.

**Comment by NotChristianGarcia** [comment]
Sorry about that! Didn't notice `build-essential` being in that docker image. I believe that was indeed the problem.

For future reference, for Docker users on windows or linux that want to mint a linux build:
- I downloaded the binary (Linux pre-built), volume mounted it and this repo into a raw ubuntu Docker image (doesn't have build-essential)
  - `docker run --rm -ti -v "UnrealEngineBinaryFolder:/UnrealEngine" -v "<UE4-SpaceMouseDir>:/UE4-SpaceMouse" ubuntu`
- Once in container I installed libudev
  - `apt update; apt install libudev-dev -y`
- Then run the build script
  - `/UnrealEngine/5.3.0/Engine/Build/BatchFiles/RunUAT.sh BuildPlugin -Plugin="/UE4-SpaceMouse/SpaceMouse.uplugin" -Package="/UE4-SpaceMouse/Out/Linux"`

* Tried in a nix-shell but I hadn't had much experience with c++ building and it was a headache, this should also work on any OS with Docker.
